python/ola/RDMTest.py

```diff
@@ -159,10 +159,66 @@ def ResponseCallback(self, response, data, unpack_exception):
     sockets[0].close()
     sockets[1].close()
 
-    self.assertTrue(results.gotrequest)
-    self.assertTrue(results.gotresponse)
+    self.assertTrue(results.got_request)
+    self.assertTrue(results.got_response)
+
+  # @timeout_decorator.timeout(2)
+  def testSetParamsWithNack(self):
+    """uses client to send an RDM set with mocked olad.
+    Regression test that confirms sent message is correct and
+    sends fixed response message."""
+    sockets = socket.socketpair()
+    wrapper = ClientWrapper(sockets[0])
+    pid_store = PidStore.GetStore(pid_store_path)
+    client = wrapper.Client()
+    rdm_api = RDMAPI(client, pid_store)
+
+    class results:
+      got_request = False
+      got_response = False
+
+    def DataCallback(self):
+      # request and response for
+      # ola_rdm_set.py -u 1 --uid 7a70:ffffff00 DMX_PERSONALITY 10
+      # against olad dummy plugin
+      # enable logging in rpc/StreamRpcChannel.py
+      data = sockets[1].recv(4096)
+      expected = binascii.unhexlify(
+        "2b000010080110001a0a52444d436f6d6d616e6422190801120908f0f401150"
+        "0ffffff180020e0012a010a30013800")
+      self.assertEqual(data, expected,
+                       msg="Regression check failed. If protocol change "
+                       "was intended set expected to: " +
+                       str(binascii.hexlify(data)))
+      results.got_request = True
+      response = binascii.unhexlify(
+        "2f0000100802100022290800100218002202000628e001300138004a0908f0f"
+        "4011500ffffff520908f0f40115ac107de05831")
+      sent_bytes = sockets[1].send(response)
+      self.assertEqual(sent_bytes, len(response))
+
+    def ResponseCallback(self, response, data, unpack_exception):
+      results.got_response = True
+      self.assertEqual(response.response_type, client.RDM_NACK_REASON)
+      self.assertEqual(response.pid, 0xe0)
```

Comment:
You should also be able to check the response.nack_reason matches (should be RDMNack.NR_DATA_OUT_OF_RANGE).

Reply:
annoyingly that isn't exposed nicely as an enum.  Think that can be fixed another time.

Reply:
Does this not work?: https://github.com/OpenLightingProject/ola/blob/master/tools/rdm/TestDefinitions.py#L1227

Reply:
I entirely missed the setattrs part.  To make myself feel better, I'll add a doc comment about it.

Reply:
Perfect thanks; brings a new meaning to self-documenting code! :laughing: